src/gui/src/ruler.cpp

```diff
@@ -0,0 +1,174 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "ord/OpenRoad.hh"
+#include "odb/db.h"
+#include "ruler.h"
+
+namespace gui {
+
+Ruler::Ruler(const odb::Point& pt0, const odb::Point& pt1, const std::string& name, const std::string& label) :
+    pt0_(pt0), pt1_(pt1), name_(name), label_(label)
+{
+  // update name if empty
+  if (name_.empty()) {
+    static int ruler_idx = 0;
+    name_ = "ruler" + std::to_string(ruler_idx++);
+  }
+}
+
+bool Ruler::operator ==(const Ruler& other) const
+{
+  return (pt0_ == other.getPt0() && pt1_ == other.getPt1()) ||
+         (pt0_ == other.getPt1() && pt1_ == other.getPt0());
+}
+
+bool Ruler::fuzzyIntersection(const odb::Point& pt, int margin)
+{
+  // compute composite line from two end points through the pt
+  // if the resultant line is longer by the margin, no intersection is reported.
+
+  auto distance = [](const odb::Point& pt0, const odb::Point& pt1) -> int {
+    const double len_x = pt0.x() - pt1.x();
+    const double len_y = pt0.y() - pt1.y();
+
+    return std::sqrt(len_x * len_x + len_y * len_y);
+  };
+
+  const int d0 = distance(pt0_, pt);
+  const int d1 = distance(pt1_, pt);
+  const int dl = distance(pt0_, pt1_);
+
+  return std::abs(d0 + d1 - dl) < margin;
+}
+
+////////////
+
+std::string RulerDescriptor::getName(std::any object) const
+{
+  auto ruler = std::any_cast<Ruler*>(object);
+  return ruler->getName();
+}
+
+std::string RulerDescriptor::getTypeName(std::any object) const
+{
+  return "Ruler";
+}
+
+bool RulerDescriptor::getBBox(std::any object, odb::Rect& bbox) const
+{
+  auto ruler = std::any_cast<Ruler*>(object);
+  bbox = odb::Rect(ruler->getPt0(), ruler->getPt1());
+  return true;
+}
+
+void RulerDescriptor::highlight(std::any object,
+               Painter& painter,
+               void* additional_data) const
+{
+  auto ruler = std::any_cast<Ruler*>(object);
+  painter.drawLine(ruler->getPt0(), ruler->getPt1());
+}
+
+Descriptor::Properties RulerDescriptor::getProperties(std::any object) const
+{
+  auto ruler = std::any_cast<Ruler*>(object);
+  const double dbu_per_uu_ = ord::OpenRoad::openRoad()->getDb()->getChip()->getBlock()->getDbUnitsPerMicron();
+  return {{"Label", ruler->getLabel()},
+          {"Point 0 - x", ruler->getPt0().x() / dbu_per_uu_},
+          {"Point 0 - y", ruler->getPt0().y() / dbu_per_uu_},
+          {"Point 1 - x", ruler->getPt1().x() / dbu_per_uu_},
+          {"Point 1 - y", ruler->getPt1().y() / dbu_per_uu_}};
+}
+
+Descriptor::Editors RulerDescriptor::getEditors(std::any object) const
+{
+  auto ruler = std::any_cast<Ruler*>(object);
+  const int dbu_per_uu_ = ord::OpenRoad::openRoad()->getDb()->getChip()->getBlock()->getDbUnitsPerMicron();
+  return {
+    {"Name", makeEditor([ruler](std::any value){
+      auto new_name = std::any_cast<const std::string>(value);
+      if (new_name.empty()) {
+        return false;
+      }
+      ruler->setName(new_name);
```

Comment:
Check for rulers with duplicate names?

Reply:
maybe the better solution would be to name the name "not editable" and use the addRuler in main window to do that check, so the uniqueness of the name cannot be modified after creation? MainWindow nevertheless holds all the rulers, so it would be much easier for it to check.

Reply:
I think we may want a find by name eventually so it would be nice to keep it editable.

Reply:
I made a ruler manager class to handle dealing with the rulers and reenabled the editing of the name.

Reply:
I would like to avoid adding singletons (aka global variables) if possible.  I think it would be cleaner to have the main window have

const std::vector<std::unique_ptr<Ruler>>& getRulers() const;

then you could have

Gui::get()->registerDescriptor<Ruler*>(new RulerDescriptor(main_window->getRulers()));

and you could then pass it into the editor lambda for dup checking.  This is closer to how the layoutViewer ctor works to access the rulers.

Reply:
No problem, I've changed it to pass in the rulers to the descriptor.